### ISSUE TITLE
Branch parsing incompatible with Git 2.4

### ISSUE THREAD

**Issue by cebamps** [Hightor/gitinfo2#14]
Git 2.4 introduced a slightly different format for decorated ref names which is incompatible with the parsing done by gitinfo2. This change in format is mentioned in the [Git 2.4.0 release notes](https://github.com/git/git/blob/master/Documentation/RelNotes/2.4.0.txt).

Here is an example of git-log "%d" formatted output:

```
 % git log -1 --decorate=short --pretty=format:"%d" HEAD
 (HEAD -> master, tag: this-is-another-tag, tag: this-is-a-tag)
```

gitinfo2 ends up mistaking refs for branch names, or if there is no ref and the decorated output is (HEAD -> master), it sets the branch name to the default `(None)`.

Although `--decorate=full` still outputs a comma in place of an arrow, [this is a bug](https://github.com/git/git/commit/5efef305d888f49ed8fce3708169a0df552ca24f) and should not be relied upon.

**Comment by longborough** [comment]
Thanks for bringing this to my attention.

Clearly, in at least one sense, this is a bug, and I'm not trying to hide from that. However, I have two problems (excuses?) to solve before I can fix it for good:
1. I only develop on Windows, and the latest git available is 1.9.5
2. The sample hooks get the data by scraping the output of git-log, Clearly, this is brittle, and I'd like to find another, stable, method for getting this data, but still within the scope of providing it as a sample hook, rather than maintaining cross-platform code to do this.

So I'm not quite sure how I'm going to fix this. Short term will probably be OK, though it will remain at the brittle mercy of changes to git-log output; long term, I simply can't answer at present. Sorry.

**Comment by mankoff** [comment]
I'm using git v 2.4.5 and I see an arrow even when using `--decorate=full`. I have fixed it (for my system) by modifying the `post-*` hooks and explicitly dealing with the branch. I haven't tested this with tags, but with a named branch it now works. The modified script follows. The changes are the three lines that contain the word **branch**:

``` bash
#!/bin/sh
# Copyright 2014 Brent Longborough
# Part of gitinfo2 package Version 2
# Please read gitinfo2.pdf for licencing and other details
# -----------------------------------------------------
# Post-{commit,checkout,merge} hook for the gitinfo2 package
#
# Get the first tag found in the history from the current HEAD
FIRSTTAG=$(git describe --tags --always --dirty='-*' 2>/dev/null)
# Get the first tag in history that looks like a Release
RELTAG=$(git describe --tags --long --always --dirty='-*' --match '[0-9]*.*' 2>/dev/null)
# Hoover up the metadata
branch=`git log -1 --decorate=short --pretty=format:"%d" HEAD`
branch=`echo $branch | sed s/\ \-\>/,/`
git --no-pager log -1 --date=short --decorate=short \
    --pretty=format:"\usepackage[%
        shash={%h},
        lhash={%H},
        authname={%an},
        authemail={%ae},
        authsdate={%ad},
        authidate={%ai},
        authudate={%at},
        commname={%an},
        commemail={%ae},
        commsdate={%ad},
        commidate={%ai},
        commudate={%at},
        refnames={${branch}},
        firsttagdescribe={$FIRSTTAG},
        reltag={$RELTAG}
    ]{gitexinfo}" HEAD > .git/gitHeadInfo.gin
```

**Comment by longborough** [comment]
Ken, thank you for that.

My plan will be to add the parsing of the "->" construct inside the package, and not inside the hooks, which I want to keep as simple as possible, since I can't be sure whether a (Windows) user would have 'extras' like `sed` available.

**Comment by TorbjornT** [comment]
Git 2.5 appears to be available for Windows now at least: https://git-scm.com/download/win

Edit: Would a simple `\StrSubstitute` in `gitexinfo.sty` do the trick, @longborough? That seemed to work here. I.e. like this:

```
\newcommand{\git@vtag}[1]{%
    \def\do##1{%
        \StrCut{##1}{tag: }\lcut\rcut%
        \IfEq{\rcut}{}{%
            \IfDecimal{\lcut}{% case where we have decimal e.g. 1.0
                \renewcommand{\gitVtag}{\lcut}
                \renewcommand{\gitVtags}{\space\lcut}
                \renewcommand{\gitVtagn}{\space\lcut}
                \listbreak
            }{}%
        }{%
            \IfDecimal\rcut{% case where we have string: decimal e.g. tag: 1.0
                \renewcommand{\gitVtag}{\rcut}
                \renewcommand{\gitVtags}{\space\rcut}
                \renewcommand{\gitVtagn}{\space\rcut}
                \listbreak
            }{}%
        }%
    }%
    \expandafter\docsvlist\expandafter{#1}%
    \StrDel{#1}{)}[\bcut]%
    \StrDel{\bcut}{(HEAD}[\bcut]%
    \StrSubstitute{\bcut}{ -> }{, }[\bcut]  % <-- this one is added
    \StrCount{\bcut}{, }[\xcut]%
    \StrCut[\xcut]{\bcut}{, }{\bcut}{\xcut}%
    \IfEq{\xcut}{}{}{%
        \renewcommand{\gitBranch}{\xcut}%
    }%
}%
```

**Comment by longborough** [comment]
Thank you very much for that suggestion; I'll incorporate itas soon as I can make time.

**Comment by longborough** [comment]
Fixed in commit 601c343 for release 2.0.5
